CI: Segmentation Faults in Github Actions

When CI is running it happens very often that a segmentation fault terminates a job. The behaviour is more or less random and often can simply be fixed by restarting all failed jobs. However it is kind of annoying because CI is far from reliable at the moment.

https://github.com/stlehmann/pyads/actions/runs/26707349297/job/78719405059?pr=525#step:6:143
```
_______________ coverage: platform linux, python 3.9.25-final-0 ________________

Name                                                                                                      Stmts   Miss  Cover
-----------------------------------------------------------------------------------------------------------------------------
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/__init__.py                         11      0   100%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/ads.py                             168     10    94%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/connection.py                      252      7    97%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/constants.py                       172      0   100%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/errorcodes.py                        3      0   100%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/errorcodes_nc.py                    14      0   100%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/filetimes.py                        12      0   100%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/pyads_ex.py                        499     47    91%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/structs.py                         127     14    89%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/symbol.py                          145      2    99%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/testserver/__init__.py               4      0   100%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/testserver/__main__.py              22     22     0%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/testserver/advanced_handler.py     221     12    95%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/testserver/basic_handler.py         87      7    92%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/testserver/handler.py               19      1    95%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/testserver/testserver.py           119      3    97%
/opt/hostedtoolcache/Python/3.9.25/x64/lib/python3.9/site-packages/pyads/utils.py                            41      3    93%
-----------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                      1916    128    93%
======================= 114 passed, 1 warning in 13.76s ========================
/home/runner/work/_temp/a48da3cb-148f-461e-988e-eeb44e57d86b.sh: line 1:  3586 Segmentation fault      (core dumped) pytest -v --cov pyads
Error: Process completed with exit code 139.
```

Any ideas are welcome.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -16,6 +16,7 @@ jobs:
 
     runs-on: ubuntu-latest
     strategy:
+      fail-fast: false
       matrix:
         python-version: ['3.9', '3.10', '3.11', '3.12', '3.13', '3.14']
 

--- a/src/pyads/pyads_ex.py
+++ b/src/pyads/pyads_ex.py
@@ -1383,7 +1383,7 @@ def adsSyncDelDeviceNotificationReqEx(
     pAmsAddr = ctypes.pointer(adr.amsAddrStruct())
     nHNotification = ctypes.c_ulong(notification_handle)
     err_code = adsSyncDelDeviceNotificationReqFct(port, pAmsAddr, nHNotification)
-    del callback_store[(adr, notification_handle)]
+    callback_store.pop((adr, notification_handle), None)
     if err_code:
         raise ADSError(err_code)
 

--- a/src/pyads/pyads_ex.py
+++ b/src/pyads/pyads_ex.py
@@ -6,6 +6,7 @@
 
 """
 from typing import Union, Callable, Any, Tuple, Type, Optional, List, Dict
+import atexit
 import ctypes
 import os
 import platform
@@ -123,6 +124,10 @@ else:  # pragma: no cover, can not test unsupported platform
     raise RuntimeError("Unsupported platform {0}.".format(sys.platform))
 
 callback_store: Dict[Tuple[AmsAddr, int], Callable[[SAmsAddr, SAdsNotificationHeader, int], None]] = dict()
+
+# Clear callback_store at interpreter shutdown before ctypes tears down the C
+# library, preventing segfaults from dangling function-pointer references.
+atexit.register(callback_store.clear)
 
 
 class ADSError(Exception):

--- a/src/pyads/pyads_ex.py
+++ b/src/pyads/pyads_ex.py
@@ -6,7 +6,6 @@
 
 """
 from typing import Union, Callable, Any, Tuple, Type, Optional, List, Dict
-import atexit
 import ctypes
 import os
 import platform
@@ -124,10 +123,6 @@ else:  # pragma: no cover, can not test unsupported platform
     raise RuntimeError("Unsupported platform {0}.".format(sys.platform))
 
 callback_store: Dict[Tuple[AmsAddr, int], Callable[[SAmsAddr, SAdsNotificationHeader, int], None]] = dict()
-
-# Clear callback_store at interpreter shutdown before ctypes tears down the C
-# library, preventing segfaults from dangling function-pointer references.
-atexit.register(callback_store.clear)
 
 
 class ADSError(Exception):

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,0 +1,5 @@
+import faulthandler
+
+# Emit a Python traceback on segfault instead of a silent crash.
+# This helps diagnose any remaining shutdown-time segfaults from the C library.
+faulthandler.enable()

